go/libraries/doltcore/sqle/dolt_query_diff_table_function.go

```diff
@@ -0,0 +1,319 @@
+	newInstance := &QueryDiffTableFunction{
+		ctx:      ctx,
+		database: database,
+	}
+	node, err := newInstance.WithExpressions(expressions...)
+	if err != nil {
+		return nil, err
+	}
+	return node, nil
+}
+
+// WithCatalog implements the sql.CatalogTableFunction interface
+func (tf *QueryDiffTableFunction) WithCatalog(c sql.Catalog) (sql.TableFunction, error) {
+	newInstance := *tf
+	newInstance.catalog = c
+	pro, ok := c.(sql.DatabaseProvider)
+	if !ok {
+		return nil, fmt.Errorf("unable to get database provider")
+	}
+	newInstance.engine = gms.NewDefault(pro)
+	err := newInstance.evalQueries()
+	if err != nil {
+		return nil, err
+	}
+	return &newInstance, nil
+}
+
+func (tf *QueryDiffTableFunction) evalQuery(query sql.Expression) (sql.Schema, sql.RowIter, error) {
+	q, err := query.Eval(tf.ctx, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+	qStr, isStr := q.(string)
+	if !isStr {
+		return nil, nil, fmt.Errorf("query must be a string, not %T", q)
+	}
+	qStr = strings.TrimSpace(qStr)
+	if !strings.HasPrefix(strings.ToLower(qStr), "select") { // TODO: allow "with?"
+		return nil, nil, fmt.Errorf("query must be a SELECT statement")
+	}
+	return tf.engine.Query(tf.ctx, qStr)
+}
+
+func (tf *QueryDiffTableFunction) evalQueries() error {
+	var err error
+	if tf.schema1, tf.rowIter1, err = tf.evalQuery(tf.query1); err != nil {
+		return err
+	}
+	if tf.schema2, tf.rowIter2, err = tf.evalQuery(tf.query2); err != nil {
+		return err
+	}
+	tf.sqlSch = append(tf.sqlSch, tf.schema1.Copy()...)
+	tf.sqlSch = append(tf.sqlSch, tf.schema2.Copy()...)
+	tf.sqlSch = append(tf.sqlSch, &sql.Column{Name: "diff_type", Type: gmstypes.Text})
+	for i := range tf.schema1 {
+		tf.sqlSch[i].Source = "query_diff"
+		tf.sqlSch[i].Name = "from_" + tf.sqlSch[i].Name
+	}
+	offset := len(tf.schema1)
+	for i := range tf.schema2 {
+		idx := offset + i
+		tf.sqlSch[idx].Source = "query_diff"
+		tf.sqlSch[idx].Name = "to_" + tf.sqlSch[idx].Name
+	}
+	return nil
+}
+
+// Database implements the sql.Databaser interface
+func (tf *QueryDiffTableFunction) Database() sql.Database {
+	return tf.database
+}
+
+// WithDatabase implements the sql.Databaser interface
+func (tf *QueryDiffTableFunction) WithDatabase(database sql.Database) (sql.Node, error) {
+	ntf := *tf
+	ntf.database = database
+	return &ntf, nil
+}
+
+// Expressions implements the sql.Expressioner interface
+func (tf *QueryDiffTableFunction) Expressions() []sql.Expression {
+	return []sql.Expression{tf.query1, tf.query2}
+}
+
+// WithExpressions implements the sql.Expressioner interface
+func (tf *QueryDiffTableFunction) WithExpressions(expression ...sql.Expression) (sql.Node, error) {
+	if len(expression) != 2 {
+		return nil, sql.ErrInvalidArgumentNumber.New(tf.Name(), "2", len(expression))
+	}
+
+	for _, expr := range expression {
+		if !expr.Resolved() {
+			return nil, ErrInvalidNonLiteralArgument.New(tf.Name(), expr.String())
+		}
+		// prepared statements resolve functions beforehand, so above check fails
```

Comment:
is this new or copied? I think expressions are going to be resolved by default now, even if not literal